dozer-api/src/cache_builder/mod.rs

```diff
@@ -1,295 +1,299 @@
 
-pub fn open_or_create_cache(
-    cache_manager: &dyn RwCacheManager,
+#[derive(Debug)]
+pub struct CacheBuilder {
+    client: InternalPipelineServiceClient<Channel>,
+    cache_manager: Arc<dyn RwCacheManager>,
     labels: Labels,
-    schema: SchemaWithIndex,
-    connections: &HashSet<String>,
-    write_options: CacheWriteOptions,
-) -> Result<Box<dyn RwCache>, CacheError> {
-    match cache_manager.open_rw_cache(
-        labels.to_non_empty_string().into_owned(),
-        labels.clone(),
-        write_options,
-    )? {
-        Some(cache) => {
-            debug_assert!(cache.get_schema() == &schema);
-            Ok(cache)
-        }
-        None => {
-            let cache = cache_manager.create_cache(
-                labels.to_non_empty_string().into_owned(),
-                labels,
-                schema.0,
-                schema.1,
-                connections,
-                write_options,
-            )?;
-            Ok(cache)
-        }
-    }
+    cache_write_options: CacheWriteOptions,
+    /// The cache that's being served.
+    serving: Arc<ArcSwap<CacheReader>>,
+    /// The cache that's being built. It must have the name of the connected log's id but may not be `serving`.
+    state: CacheBuilderState,
+    log_reader_options: LogReaderOptions,
+    log_reader: LogReader,
 }
 
-const READ_LOG_RETRY_INTERVAL: Duration = Duration::from_secs(1);
+mod endpoint_meta;
+mod state;
 
-async fn read_log_task(
-    mut cancel: impl Future<Output = ()> + Unpin + Send + 'static,
-    mut log_reader: LogReader,
-    sender: mpsc::Sender<OpAndPos>,
-) {
-    loop {
-        let next_op = std::pin::pin!(log_reader.read_one());
-        match select(cancel, next_op).await {
-            Either::Left(_) => break,
-            Either::Right((op, c)) => {
-                let op = match op {
-                    Ok(op) => op,
-                    Err(e) => {
-                        error!(
-                            "Failed to read log: {e}, retrying after {READ_LOG_RETRY_INTERVAL:?}"
-                        );
-                        tokio::time::sleep(READ_LOG_RETRY_INTERVAL).await;
-                        cancel = c;
-                        continue;
-                    }
-                };
+use endpoint_meta::EndpointMeta;
+use state::CacheBuilderState;
 
-                cancel = c;
-                if sender.send(op).await.is_err() {
-                    debug!("Stop reading log because receiver is dropped");
-                    break;
-                }
-            }
-        }
-    }
-}
+impl CacheBuilder {
+    pub async fn new(
+        cache_manager: Arc<dyn RwCacheManager>,
+        app_server_url: String,
+        endpoint: &ApiEndpoint,
+        labels: LabelsAndProgress,
+    ) -> Result<(Self, Vec<u8>), ApiInitError> {
+        // Connect to the endpoint's log.
+        let mut client = InternalPipelineServiceClient::connect(app_server_url.clone())
```

Comment:
Should we try to re-use the (re-)connect logic? We can use the cancellation future for time-outs.

Reply:
I'm not sure how to reuse reconnect code. We don't recreate `InternalPipelineServiceClient` now (or should we?). For now the reconnect happens inside tonic. We just call `InternalPipelineServiceClient::get_id` and etc again on network failure.